Device Binding Callback implementation

vahancouver opened the pull request:
# JIRA Ticket

[SDKS-1748](https://bugster.forgerock.org/jira/browse/SDKS-1748)

# Description

Device Binding callback implementation
FRUI changes for testing in the example app

Note: tests will be implemented as a separate ticket

# Definition of Done Checklist:

- [ ] Acceptance criteria is met.
- [ ] All tasks listed in the user story have been completed.
- [ ] Coded to standards.
- [ ] Ensure backward compatibility.
- [ ] API reference docs is updated.
- [ ] Unit tests are written.
- [ ] Integration tests are written.
- [ ] e2e tests are written.
- [ ] Functional spec is written/updated.
- [ ] Example code snippets have been added.
- [ ] Change log updated.
- [ ] Documentation story is created and tracked.
- [ ] Tech debts and remaining tasks are tracked in separated ticket(s).

jeyanthanperiyasamy commented on FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift:
```diff
@@ -0,0 +1,221 @@
+            throw AuthError.invalidCallbackResponse(String(describing: json))
+        }
+        
+        for output in outputs {
```
`can we refactor this code ? we could do some thing like this

   inputs.forEach { input in
            foo(input[CBConstants.name] as? String)
        }
        
        `func` foo(name: String) {
        switch name {
        case Constants.jws: {
            self.jws = name
        }
        }
    }
`

rodrigoareis commented on FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift:
```diff
@@ -0,0 +1,221 @@
+//
+
+
+import Foundation
```
Add new line separator

rodrigoareis commented on FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift:
```diff
@@ -0,0 +1,266 @@
+import JOSESwift
+
+
+/// Protiocol  to override keypair generation, authentication, signing and access control
```
Typo issue

rodrigoareis commented on FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift:
```diff
@@ -0,0 +1,266 @@
+    /// Initializer for BiometricOnly
+    /// - Parameter biometricInterfcae: biometric handler for autentication
+    /// - Parameter keyAware: keyAware for key pair generation
+    init(biometricInterfcae: BiometricHandler, keyAware: KeyAware) {
```
Typo issue. `biometricInterface`

jeyanthanperiyasamy reviewed (changes requested):
not done yet..will review sometime later

rodrigoareis commented on FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift:
```diff
@@ -0,0 +1,221 @@
+            throw AuthError.invalidCallbackResponse(String(describing: json))
+        }
+        
+        for output in outputs {
```
Do you think we can simplify this interaction logic?

jeyanthanperiyasamy replied:
+1

jeyanthanperiyasamy replied:
my idea is something like this. 

    inputs.forEach { input in
            foo(input[CBConstants.name] as? String)
        }

 `    func foo(name: String) {
        switch name {
        case Constants.jws: {
            self.jws = name
        }
       case Constants.authenticationType: {
            self.deviceBindingAuthenticationType = DeviceBindingAuthenticationType(rawValue: name)
        }
        }
    }`

rodrigoareis commented on FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift:
```diff
@@ -0,0 +1,221 @@
+            throw AuthError.invalidCallbackResponse("Missing an output value")
+        }
+        
+        for input in inputs {
```
Do you think we can simplify this interaction logic?

rodrigoareis reviewed (changes requested):
Left minor comments to be reviewed. Is there any kind of tests for new `DeviceBindingCallback`?

jeyanthanperiyasamy commented on FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift:
```diff
@@ -0,0 +1,221 @@
+    /// - Parameter authInterface: Interface to find the Authentication Type - procide nil to default to getDeviceBindingAuthenticator()
+    /// - Parameter deviceId: Interface to find the Authentication Type - provide nil to default to FRDevice.currentDevice?.identifier.getIdentifier()
+    /// - Parameter completion: completion Completion block for Device binding result callback
+    open func execute(authInterface: DeviceAuthenticator?,
```
i am not sure we need open,  protected open is a standard in kotlin to override and method is not exposed , in swift open method can be called by anyone i believe . so i am kind of thinking this

`class DeviceBindingCallback {
    fileprivate func execute() {
        print("Foo")
    }
}

class Bar: DeviceBindingCallback {
    fileprivate override func jey() {
        print("Bar")
    }
}`

jeyanthanperiyasamy replied:
so i dont want this 

let callback = DeviceBindCallBack()
callback.execute()

but i want this

class GeorgeCallback: DeviceBindCallBack {
   override func execute()
}

jeyanthanperiyasamy commented on FRAuth/FRAuth/DeviceBinding/BiometricBindingHandler.swift:
```diff
@@ -0,0 +1,100 @@
+    
+    /// Display biometric prompt  for Biometric and device credential if needed
+    /// - Parameter timeout: Timeout for the biometric prompt
+    /// - Parameter completion: Completion block for Device binding result callback
```
format this class , looks like its not formatted

jeyanthanperiyasamy commented on FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift:
```diff
@@ -0,0 +1,266 @@
+    /// - Returns: compact serialized jws
+    func sign(keyPair: KeyPair, kid: String, userId: String, challenge: String) throws -> String {
+        
+        let jwk = try RSAPublicKey(publicKey: keyPair.publicKey, additionalParameters: ["use": "sig", "alg": "RS512"])
```
we need to create a keys using secure enclave ?

jeyanthanperiyasamy replied:
please check they support RSA/SHA and let us know . Andy may have to do some work on server if thats the only thing iOS supports

jeyanthanperiyasamy commented on FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift:
```diff
@@ -0,0 +1,221 @@
+                self.deviceIdKey = name
+            }
+            else if let name = input[CBConstants.name] as? String, name.contains(CBConstants.clientError) {
+                self.clientErrorKey = name
```
For consistency, do we need self.jwsKey = name or just jwsKey = name, if the global variable and local variable same , people prefer self.

jeyanthanperiyasamy commented on FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift:
```diff
@@ -0,0 +1,221 @@
+    
+    /// Handle all the errors for the device binding.
+    /// - Parameter status: Device binding status
+    /// - Parameter completion: Completion block Device binding result callback
```
open should not be exposed, same comment as above

george-bafaloukas-forgerock commented on FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift:
```diff
@@ -0,0 +1,221 @@
+    /// The authentication type of the journey
+    private(set) var deviceBindingAuthenticationType: DeviceBindingAuthenticationType!
+    // The title to be displayed in biometric prompt
+    private(set) var title: String!
```
Should that be forced?
Can we not have them without the "!" either some as optional or the rest set with a default value or on the init

george-bafaloukas-forgerock commented on FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift:
```diff
@@ -0,0 +1,266 @@
+
+
+
+
```
There are some formatting issues on this file, double spaces and extra lines

george-bafaloukas-forgerock commented on FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift:
```diff
@@ -0,0 +1,221 @@
+            return
+        }
+        
+        newAuthInterface.authenticate(timeout: timeout ?? 60) { result in
```
Are we capturing self on this call? might be work changing that to 
"newAuthInterface.authenticate(timeout: timeout ?? 60) {  [unowned self] result in"

And test if we are OK

jeyanthanperiyasamy commented on FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift:
```diff
@@ -0,0 +1,221 @@
+    
+    
+    /// Bind the device.
+    /// - Parameter completion Completion block for Device binding result callback
```
we want this to be open func bind()

spetrov commented on FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift:
```diff
@@ -0,0 +1,266 @@
+        header.jwkTyped = jwkWithKeyId
+        
+        //create payload
+        let params: [String: Any] = ["sub": userId, "challenge": challenge, "exp": (Int(Date().addingTimeInterval(30).timeIntervalSince1970))]
```
The `exp` value should be set with the value from the `timeout` attribute coming with the callback. 
In android, the following logic is in place:

>   /**
>      * Get Expiration date for the signed token, claim "exp" will be set to the JWS.
>      *
>      * @return The expiration date
>      */
>     protected open fun getExpiration(): Date {
>         val date = Calendar.getInstance();
>         date.add(Calendar.SECOND, timeout ?: 60)
>         return date.time;
>     }

spetrov commented on FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift:
```diff
@@ -0,0 +1,266 @@
+        
+        //create signer
+        guard let signer = Signer(signingAlgorithm: .RS512, key: keyPair.privateKey) else {
+            throw DeviceBindingStatus.unsupported(errorMessage: "Cannot creat a signer for jws")
```
Typo in `errorMessage`: "creat" => "create"

spetrov commented on FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift:
```diff
@@ -0,0 +1,266 @@
+/// DeviceAuthenticator adoption for  biometric only authentication
+internal struct BiometricOnly: DeviceAuthenticator {
+    
+    /// biometric handler for autentication
```
Typo: "autentication" => "authentication"

spetrov commented on FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift:
```diff
@@ -0,0 +1,266 @@
+}
+
+
+/// DeviceAuthenticator adoption for  biometric only authentication
```
Remove extra space between "for" and "biometric"

spetrov commented on FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift:
```diff
@@ -0,0 +1,266 @@
+    }
+    
+    
+    /// Generate  public and private key pair
```
Remove extra space between "Generate" and "public"

spetrov commented on FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift:
```diff
@@ -0,0 +1,266 @@
+    
+    
+    /// Display authentication  prompt for authentication type if needed
+    /// - Parameter timeout: Timeout for teh authentication prompt
```
Typo: "teh"...
Remove extra space in "authentication  prompt"

spetrov reviewed (changes requested):
Reviewed. I have a few minor issues that we may need to address...

jeyanthanperiyasamy commented on FRAuth/FRAuth/DeviceBinding/KeyAware.swift:
```diff
@@ -0,0 +1,99 @@
+        
+        query[String(kSecAttrKeyType)] = String(keyType)
+        query[String(kSecAttrKeySizeInBits)] = keySize
+        
```
i belive the secure enclave the part will be somewhere here 

` #if !targetEnvironment(simulator)
        // If the device supports Secure Enclave, create a keypair using Secure Enclave TokenID
        if SecuredKey.isAvailable() {
            query[String(kSecAttrTokenID)] = String(kSecAttrTokenIDSecureEnclave)
            let accessControl = SecAccessControlCreateWithFlags(kCFAllocatorDefault, accessibility.rawValue as CFString, [.privateKeyUsage, .userPresence] , nil)!
            keyAttr[String(kSecAttrAccessControl)] = accessControl
        }
        #endif`

vahancouver commented:
@spetrov @jeyanthanperiyasamy @rodrigoareis @george-bafaloukas-forgerock Please have another look. 

I addressed all the comments related to formatting, double spaces and typos.

I have changed to use the Secure Enclave if available and the token is generated now using ES256 algorithm.

@george-bafaloukas-forgerock @jeyanthanperiyasamy let's connect early morning my time to go over the parsing code in DeviceBindingCallback.init method and the reason of using forced unwrapped properties. I made some changes here to use switch [though arguably I preferred the previous code as it was more readable]

jeyanthanperiyasamy commented on FRAuth/FRAuthTests/FRAuthSwiftTests/FRAuth/Cookie/CookieValidationTests.swift:
```diff
@@ -202,7 +202,7 @@ class CookieValidationTests: FRAuthBaseTest {
     
     func test_10_cookie_is_expired_validation_not_expired() {
-        let setCookie: [String: String] = ["Set-Cookie":"iPlanetDirectoryPro=token; Expires=Wed, 21 Oct 2022 01:00:00 GMT; Domain=openam.example.com"]
+        let setCookie: [String: String] = ["Set-Cookie":"iPlanetDirectoryPro=token; Expires=Wed, 21 Oct 2032 01:00:00 GMT; Domain=openam.example.com"]
```
is this part of this story ? or its a accident commit

jeyanthanperiyasamy commented on FRTestHost/FRTestHost/FRAuthConfigPKHash.plist:
```diff
@@ -26,7 +26,7 @@
 	<key>forgerock_ssl_pinning_public_key_hashes</key>
 	<array>
-		<string>+KSzREQbAh9gqYLLGpfCG+cAy7Px3/Qmk/e8Egwyd7o=</string>
+		<string>9uiEUVtJyf/LAciFcfkWlBcrPor9uZGmteP3x9sBGiQ=</string>
```
is this part of this story ? or its a accident commit

jeyanthanperiyasamy reviewed (changes requested):
most of the changes looks good now, Lets test the ECC encryption type , once and wrap this story

vahancouver commented:
@jeyanthanperiyasamy @spetrov 
Turns out the system triggers (biometric) authentication on signing operation when a private key is used that was created in secure enclave with the appropriate access flags. 
Because of this for the BIOMETRIC_ONLY and BIOMETRIC_ALLOW_FALLBACK types we had double authentication (I was triggering the second one in the code explicitly). 
With the latest changes I removed the manual triggering of the authentication. Now it gets triggered only once by the system during the JWS signing operation. 
Please review my changes.

jeyanthanperiyasamy commented on FRAuth/FRAuth/DeviceBinding/KeyAware.swift:
```diff
@@ -0,0 +1,102 @@
+
+///Helper struct to generate and sign the keys
+struct KeyAware {
+    var userId: String
```
userid and time is suppose to be private set and public get ?

jeyanthanperiyasamy commented on FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift:
```diff
@@ -0,0 +1,254 @@
+    
+    //  MARK: - Properties
+    
+    /// The userId received from server
```
can you check the Android changes for this , all this needs to be private set and public get i belive ?

jeyanthanperiyasamy commented on FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift:
```diff
@@ -0,0 +1,250 @@
+    
+    
+    /// Access Control for the authetication type
+    func accessControl() -> SecAccessControl? {
```
lets discuss about this fallback mechanism

jeyanthanperiyasamy commented on FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift:
```diff
@@ -0,0 +1,250 @@
+    
+    /// Access Control for the authetication type
+    func accessControl() -> SecAccessControl? {
+#if !targetEnvironment(simulator)
```
same we need to discuss about this

jeyanthanperiyasamy commented on FRAuth/FRAuth/DeviceBinding/KeyAware.swift:
```diff
@@ -0,0 +1,102 @@
+        keyAttr[String(kSecAttrApplicationTag)] = key
+        query[String(kSecPrivateKeyAttrs)] = keyAttr
+        
+#if !targetEnvironment(simulator)
```
if the secure key is not available , should we fail or guide the users ? something to think of here

jeyanthanperiyasamy reviewed (changes requested):
feedback given

rodrigoareis reviewed (approved):
Changes looks good to me

jeyanthanperiyasamy commented on FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift:
```diff
@@ -0,0 +1,250 @@
+    func accessControl() -> SecAccessControl? {
+#if !targetEnvironment(simulator)
+        return SecAccessControlCreateWithFlags(kCFAllocatorDefault, kSecAttrAccessibleWhenPasscodeSetThisDeviceOnly, SecuredKey.isAvailable() ? [.userPresence, .privateKeyUsage] : [.userPresence], nil)!
+#else
```
i am thinking like if its debug mode use this  [.userPresence] and for release [.userPresence, .privateKeyUsage] ...so we can remove the SecureEnclave check ...thoughts ?

jeyanthanperiyasamy commented on FRAuth/FRAuth/DeviceBinding/KeyAware.swift:
```diff
@@ -0,0 +1,114 @@
+        
+        var error: Unmanaged<CFError>?
+        guard let privateKey = SecKeyCreateRandomKey(builderQuery as CFDictionary, &error) else {
+            throw error!.takeRetainedValue() as Error
```
can this be optional   - error?.takeRetainedValue()

jeyanthanperiyasamy commented on FRAuth/FRAuth/DeviceBinding/KeyAware.swift:
```diff
@@ -0,0 +1,114 @@
+        guard status == errSecSuccess else {
+            return nil
+        }
+        return (item as! SecKey)
```
can this be optional - return (item as? SecKey)     ?

jeyanthanperiyasamy commented on FRAuth/FRAuth/DeviceBinding/KeyAware.swift:
```diff
@@ -0,0 +1,114 @@
+    /// - Parameter keyName: key names to be hashed
+    /// - Returns: the hash for the given key name
+    static func getKeyAlias(keyName: String) -> String {
+        let data = keyName.data(using: .utf8)!
```
same here...can we do without bangs

jeyanthanperiyasamy commented on FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift:
```diff
@@ -0,0 +1,250 @@
+    func accessControl() -> SecAccessControl? {
+#if !targetEnvironment(simulator)
+        return SecAccessControlCreateWithFlags(kCFAllocatorDefault, kSecAttrAccessibleWhenPasscodeSetThisDeviceOnly, SecuredKey.isAvailable() ? [.biometryCurrentSet, .privateKeyUsage] : [.biometryCurrentSet], nil)!
+#else
```
do we need this check and bang - SecuredKey.isAvailable()

jeyanthanperiyasamy reviewed (changes requested):
verify what comments we can address
